nxcomp/src/Timestamp.cpp

```diff
@@ -44,34 +44,57 @@
 
 T_timestamp timestamp;
 
-//
-// The following functions all use the ctime
-// static buffer from the C library.
-//
-
-char *strTimestamp(const T_timestamp &ts)
+std::string strTimestamp(const T_timestamp &ts)
 {
-  char *ctime_now = ctime((time_t *) &ts.tv_sec);
+  std::string ret;
+
+  char ctime_now[26] = { };
+  bool err = true;
+
+#if HAVE_CTIME_S
+  errno_t retval = ::ctime_s(ctime_now, 26, static_cast<const time_t*>(&ts.tv_sec));
 
-  ctime_now[24] = '\0';
+  err = (retval != 0);
+#else
+  char *retval = ::ctime_r(static_cast<const time_t*>(&ts.tv_sec), ctime_now);
 
-  return ctime_now;
+  err = (!(retval));
+#endif
+
+  if (err)
+  {
+    std::cerr << "WARNING: converting time to string failed." << std::endl;
+  }
+  else
+  {
+    /* Replace newline at position 25 with a NULL byte. */
+    ctime_now[24] = '\0';
+
+    ret = ctime_now;
+  }
+
+  return ret;
 }
 
 //
-// This is especially dirty. 
+// This is especially dirty.
 //
 
-char *strMsTimestamp(const T_timestamp &ts)
+std::string strMsTimestamp(const T_timestamp &ts)
 {
-  char *ctime_now = ctime((time_t *) &ts.tv_sec);
+  std::string ret;
+
+  std::string ctime_now = strTimestamp(ts);
 
-  char ctime_new[25];
+  if (!(ctime_now.empty()))
+  {
+    char ctime_new[26] = { };
 
-  sprintf(ctime_new, "%.8s:%3.3f", ctime_now + 11,
-              (float) ts.tv_usec / 1000);
+    snprintf(ctime_new, sizeof(ctime_new), "%.8s:%3.3f",
+             ctime_now.c_str() + 11, static_cast<float>(ts.tv_usec) / 1000);
 
-  strncpy(ctime_now, ctime_new, 24);
+    ret = ctime_new;
+  }
 
-  return ctime_now;
+  return ret;
```

Comment:
One thing: isn't there a timestamp class in C++ that will generate what we need? We could drop the C function like snprintf altogether.

Reply:
ping?

Reply:
*sigh* No, and I probably have to explain this.

The C++ standard doesn't include support for calendric services. Support is limited to timers at best, but nothing that would need a calender. Implementing stuff like this is a snake pit and the standards committee didn't want to delve into this only to in the end find out that it's unusable anyway since tiny details have been missed. So to save a great amount of time and spare themselves headaches, they decided to drop that matter completely.

And so they did till (almost) this day.

For such things, the only usable alternative are C functions.

The `ctime` function family is deprecated in favor of `strftime` (and especially the `ctime()` function not even thread-safe), but while implementing the new logging code that used `strftime` in the beginning I noticed that `strftime` is not implemented in a thread-safe fashion in glibc (although that's no standards problem, since it COULD be thread-safe - glibc just adds stuff that breaks that).

The only viable alternative would be `std::put_time`, which should be analogous to `strftime`, but maybe better implemented. This function is only available since C++11 though, and we cannot universally depend upon that (i.e., there are still platforms we want to build on that don't support this standard).

For now it's probably best to stay with `ctime_r`/`ctime_s`.

Reply:
ok, thx.